### PR TITLE
Refiner refactor

### PR DESCRIPTION
**Pull request by dagewa**
This branch contains changes related to tidying and refactoring DIALS refinement code, particularly RefinerFactory. Individual components used by a Refiner, such as the ExperimentsPredictor, Target, ReflectionManager, and so on now have their factory code and PHIL definitions defined in their own modules, rather than all in one place in refiner.py, improving encapsulation. Moving some code from Target to ExperimentsPredictor allows construction of objects to occur in a different order and ensures that a Target is built only when everything else is ready. Configuring the parameterisation is the most involved part of the refiner construction. Now this has been broken down so that the parameterisation auto reduction options are handled by a separate class.

There are numerous other changes refactoring classes, tidying imports, changing method names to be more descriptive, improving docstrings etc. The intention is that there will be no effect on the Refiner itself - refinement results should not change between this branch and master, so this is purely housekeeping. All DIALS tests pass, however a review is requested from @phyy-nx, as there are XFEL tests that may provide more coverage of advanced use cases than the DIALS tests (adding these to automatically-run DIALS tests is another issue to be addressed).

More housekeeping could be done, but I'm keen to merge this soon to avoid diverging further from master and conflicts with others' developments. Some of the commits could be squashed on merge.

### CODE REVIEW

**Comment by dagewa** [comment]
Thanks @phyy-nx! I agree it would be good to get the ensemble refinement included in tests. As the data for that will not change it might be better not to put them under svn in dials_regression, but to host them externally somewhere (zenodo?) and allow them to be included in `dials.fetch_test_data`. This is the way xia2 regression tests have gone. @Anthchirp has the details. I'm not sure what is being considered for DIALS regression tests, but it would be a worthy discussion topic.

File: algorithms/refinement/target.py [hunk 1]
```diff
@@ -18,9 +18,92 @@
...
+    # restraints_parameterisation, as these are not required for initialisation
+    # and will be linked to the object later
+    return targ(experiments=experiments,
+                predictor=predictor,
+                reflection_manager=reflection_manager,
+                prediction_parameterisation=pred_param,
+                restraints_parameterisation=restraints_param,
+                frac_binsize_cutoff=params.bin_size_fraction,
+                absolute_cutoffs=absolute_cutoffs,
+                gradient_calculation_blocksize=params.gradient_calculation_blocksize)
```
[the remark this answers is not in the document]
Comment (dagewa) [reply algorithms/refinement/target.py]: Yes, @ndevenish took exception to these as well. Loosely, the class is being used like a namespace, so `TargetFactory.from_parameters_and_experiments` resembles e.g. `ExperimentListFactory.from_json_file`. In the latter case it makes more sense though as there are multiple factory functions under the class. I would like to leave this as-is for now.

File: algorithms/refinement/reflection_manager.py [hunk 2]
```diff
@@ -141,6 +214,106 @@ def per_image(self):
...
+        params The input parameters
+
+    Returns:
+        The reflection manager instance
+    """
+
+    # While a random subset of reflections is used, continue to
+    # set random.seed to get consistent behaviour
+    if params.random_seed is not None:
+      import random
```
[the remark this answers is not in the document]
Comment (Anthchirp) [reply algorithms/refinement/reflection_manager.py]: Yes. An import of a standard python library should almost always go to the top of the file.

File: algorithms/refinement/parameterisation/autoreduce.py [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+    free_gp_params = [a and not b for a,b in zip(gp_params, fixlist)]
+    nparam = free_gp_params.count(True)
+    cutoff = self._options.min_nref_per_parameter * nparam
+    isel = flex.size_t()
+    surplus = pg_surpl(self.reflections["id"], self.reflections["panel"], pnl_ids,
+        exp_ids, cutoff).result
+
+    if surplus < 0 and verbose:
+      logger.warning('{0} reflections on panels {1} '
+                     'with a cutoff of {2}'.format(nref, pnl_ids, cutoff))
```
[the remark this answers is not in the document]
Comment (dagewa) [reply algorithms/refinement/parameterisation/autoreduce.py]: It is worse than that, even in master the definition of nref got [commented out](https://github.com/dials/dials/blob/master/algorithms/refinement/refiner.py#L1262). I conclude that this never gets called, so I'll just remove it

**Comment by dagewa** [comment]
@Anthchirp thanks for the comments. I've pushed fixes that resolve most of these, I think.

**Comment by dagewa** [comment]
ah, I did not see new comments till now (had not reloaded the page). I'll look at these tomorrow

File: algorithms/refinement/parameterisation/configure.py [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+    sv_xl_uc = options.scan_varying and not options.crystal.unit_cell.force_static
+    for icrystal, crystal in enumerate(experiments.crystals()):
+      # This crystal can only ever appear either in scans or in stills
+      # (otherwise it requires a different crystal model)
+      exp_ids = experiments.indices(crystal)
+      assoc_models = [(experiments[i].goniometer, experiments[i].scan) \
+                      for i in exp_ids]
+      goniometer, scan = assoc_models[0]
+      if goniometer is None:
+        if not all(g is None and s is None for (g, s) in assoc_models):
```
[the remark this answers is not in the document]
Comment (dagewa) [reply algorithms/refinement/parameterisation/configure.py]: Yes, that's clearer

File: algorithms/refinement/parameterisation/configure.py [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+      # get number of fixable units, either parameters or parameter sets in
+      # the scan-varying case
+      try:
+        num_ori = xl_ori_param.num_sets()
+      except AttributeError:
+        num_ori = xl_ori_param.num_total()
+      try:
+        num_uc = xl_uc_param.num_sets()
+      except AttributeError:
+        num_uc = xl_uc_param.num_total()
```
[the remark this answers is not in the document]
Comment (dagewa) [reply algorithms/refinement/parameterisation/configure.py]: I considered this alternative

```python
get_num_uc = getattr(xl_uc_param, 'num_sets', None)
if not get_num_uc:
  get_num_uc = getattr(xl_uc_param, 'num_total')
num_uc = get_num_uc()
```

but I've gone with yours as the assignment to `None` and then test of that is extra work for merely cosmetic reasons.

File: algorithms/refinement/parameterisation/configure.py [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+          from .prediction_parameters_stills \
+            import StillsPredictionParameterisationSparse as spp
+      else:
+        if options.spherical_relp_model:
+          from .prediction_parameters_stills \
+            import SphericalRelpStillsPredictionParameterisation as spp
+        else:
+          from .prediction_parameters_stills \
+              import StillsPredictionParameterisation as spp
+      pred_param = spp(experiments, det_params, beam_params, xl_ori_params,
```
[the remark this answers is not in the document]
Comment (dagewa) [reply algorithms/refinement/parameterisation/configure.py]: Hmm, yes. I went down the wrong route long ago with naming, ending up with descriptive, but very long, class names and a deep package/module hierarchy. I aimed to keep the distinction between models and their parameterisations obviously clear, but I agree once in the `parameterisation/` package there isn't a need to repeat `parameterisation` in every module or class name.

I'll remove the inline imports for now and will have to think further about a wider redesign for later. The Strategy pattern seems appropriate. I need it in a few places during construction of the Refiner where choices of particular classes are made depending on user input and the data. It was always a mess of conditional statements. I had been looking at creational patterns for complex objects to get around this but hadn't considered Strategy.

File: algorithms/refinement/parameterisation/configure.py [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+    # Use the results of centroid analysis to suggest suitable interval widths
+    # for each experiment. This will be the smallest of the proposed intervals
+    # for each of the residuals in x, y and phi, as long as this is not smaller
+    # than either the outlier rejection block width, or 9.0 degrees.
+    for i, a in enumerate(analysis):
+      intervals = [a.get('x_interval'),
+                   a.get('y_interval'),
+                   a.get('phi_interval')]
+      try:
+        min_interval = min([e for e in intervals if e is not None])
```
[the remark this answers is not in the document]
Comment (dagewa) [reply algorithms/refinement/parameterisation/configure.py]: Oh, I did not know that.

**Comment by dagewa** [comment]
@Anthchirp Ok, now I think I've dealt with most of the above, accepting there could be more to do in a broader redesign for a later version.

File: algorithms/refinement/parameterisation/configure.py [hunk 8]
```diff
@@ -491,11 +482,12 @@ def from_parameters_and_experiments(cls, options, experiments,
       else:
         if options.detector.panels == "automatic":
           if len(detector) > 1:
-            try: # Use hierarchy in parameterisation if the detector has one
+            if hasattr(detector, 'hierarchy'):
+              # Use hierarchy in parameterisation if the detector has one
               h = detector.hierarchy()
```
Comment (Anthchirp) [inline algorithms/refinement/parameterisation/configure.py]: Suspect this line can now be removed

**Comment by Anthchirp** [comment]
Thank you for addressing the comments. At currently 35 commits would suggest either
* squash merging (1 commit in project history, but original commits will remain in the pull request),
 or 
* drastically reducing the number of commits in the pull request (_n_ commits in project history, but original commits will not be directly accessible)

File: algorithms/refinement/parameterisation/configure.py (hunk 8 above)
Comment (dagewa) [reply algorithms/refinement/parameterisation/configure.py]: oops, yes

**Comment by dagewa** [comment]
Hmm... decisions. Given that refinement results should not have changed (and the tests support that) then squashing to a single commit seems reasonable. However, how long will the branch be available for? I mean, I guess forever locally if I don't delete my copy, but if we discover some bug in a year's time introduced by one of these changes, will the branch still be accessible on GitHub to debug the issue?

Another issue is that @phyy-nx  has changes he would like to merge in. That might be easier with a history of incremental commits. What do you think?

**Comment by phyy-nx** [comment]
On my end, it's not an issue. I won't be merging in for a while and I
expect to fall behind.

**Comment by dagewa** [comment]
Thanks. I've decided I'd like to keep a few separate commits in the history. I'll talk to @Anthchirp about rewriting the pull request tomorrow

**Pull request closed without merging**